mm2src/coins/hd_wallet/mod.rs

```diff
@@ -0,0 +1,431 @@
+    /// receives transactions, they won't be identified.
+    pub gap_limit: u32,
+}
+
+#[async_trait]
+impl<HDAccount> HDWalletOps for HDWallet<HDAccount>
+where
+    HDAccount: HDAccountOps + Clone + Send + Sync,
+{
+    type HDAccount = HDAccount;
+
+    fn coin_type(&self) -> u32 { self.derivation_path.coin_type() }
+
+    fn gap_limit(&self) -> u32 { self.gap_limit }
+
+    fn account_limit(&self) -> u32 { DEFAULT_ACCOUNT_LIMIT }
+
+    fn default_receiver_chain(&self) -> Bip44Chain { DEFAULT_RECEIVER_CHAIN }
+
+    fn get_accounts_mutex(&self) -> &HDAccountsMutex<Self::HDAccount> { &self.accounts }
+
+    async fn get_account(&self, account_id: u32) -> Option<Self::HDAccount> {
+        let accounts = self.get_accounts_mutex().lock().await;
+        accounts.get(&account_id).cloned()
+    }
+
+    async fn get_account_mut(&self, account_id: u32) -> Option<HDAccountMut<'_, Self::HDAccount>> {
+        let accounts = self.get_accounts_mutex().lock().await;
+        if !accounts.contains_key(&account_id) {
+            return None;
+        }
+
+        Some(AsyncMutexGuard::map(accounts, |accounts| {
+            accounts
+                .get_mut(&account_id)
+                .expect("getting an element should never fail due to the checks above")
+        }))
+    }
+
+    async fn get_accounts(&self) -> HDAccountsMap<Self::HDAccount> { self.get_accounts_mutex().lock().await.clone() }
+
+    async fn get_accounts_mut(&self) -> HDAccountsMut<'_, Self::HDAccount> { self.get_accounts_mutex().lock().await }
+
+    async fn remove_account_if_last(&self, account_id: u32) -> Option<Self::HDAccount> {
+        let mut x = self.get_accounts_mutex().lock().await;
+        // `BTreeMap::last_entry` is still unstable.
+        let (last_account_id, _) = x.iter().last()?;
+        if *last_account_id == account_id {
+            x.remove(&account_id)
+        } else {
+            None
+        }
+    }
+
+    async fn get_enabled_address(&self) -> Option<HDWalletAddress<Self>> {
+        let enabled_address = match self.enabled_address.clone() {
+            Some(id) => id,
+            None => return None,
+        };
+
+        let account = match self.get_account(enabled_address.account_id).await {
+            Some(account) => account,
+            None => return None,
+        };
+
+        let hd_address_id = HDAddressId {
+            chain: enabled_address.chain,
+            address_id: enabled_address.address_id,
+        };
+
+        let address = account
+            .derived_addresses()
+            .lock()
+            .await
+            .get(&hd_address_id)
+            .map(|addr| addr.address());
+
+        address
+    }
+}
+
+#[async_trait]
+impl<HDAccount> HDWalletStorageOps for HDWallet<HDAccount>
+where
+    HDAccount: HDAccountOps + Clone + Send + Sync,
+{
+    fn hd_wallet_storage(&self) -> &HDWalletCoinStorage { &self.hd_wallet_storage }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
```

Comment:
`Copy` can also be derived for `HDAccountAddressId`.

Reply:
> `Copy` can also be derived for `HDAccountAddressId`.

I have a question, why there is a need in `Copy` for  this struct?

Reply:
> Generally speaking, if your type can implement Copy, it should.

https://doc.rust-lang.org/std/marker/trait.Copy.html#when-should-my-type-be-copy

🙂

Reply:
Interesting, ok :D

Reply:
done